### PR TITLE
Rholang: Add initial work on normalizer.

### PR DESCRIPTION
**Pull request by guardbotmk3**
Normalizer will work in 2 parts:
Part 1 will "compile" the source down to a partially normalized form, and part 2
will sort the subcomponents of all subproccesses according to an order.

This is the beginning of part 1: it adds some visitor classes to start producing
a normalized form from the syntax, and tests! to make sure that that the few
productions for which we have visitors are working.

Adds scalatest as a testing dependency.

### CODE REVIEW

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Normalization:
+// Normalization requires the evaluation of everything that doesn't go through
+// the tuplespace, so the top level of a normalized process must have everything
+
+package coop.rchain.interpreter
+
+case class Par(
+  sends: List[Send],
+  receives: List[Receive],
```
Comment (pyrocto) [inline rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: I presume contracts are classified as receives and invocations as sends?

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// In the DeBruijn level paper, they use negatives, but this is more clear.
+case class FreeVar(level: Int) extends Var
+
+// upon send, all free variables in data are substituted with their values.
+// also if a process is sent, it is auto-quoted.
+case class Send(chan: Channel, data: Channel)
+
+// [Par] is an n-arity Pattern.
+// It's an error for free Variable in the list of patterns to overlap.
```
Comment (pyrocto) [inline rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: This comment is hard to parse as written.  I think you mean that it's an error for a single free variable to occur in more than one place in a pattern.

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+case class BoundVar(level: Int) extends Var
+// It is an error to use WildCard() in a non-binding position
+case class WildCard() extends Var
+// Variables that occur free in Par used as a pattern or ChanVar are binders.
+// For the purpose of comparing patterns, we count just like BoundVars.
+
+// In the DeBruijn level paper, they use negatives, but this is more clear.
+case class FreeVar(level: Int) extends Var
+
+// upon send, all free variables in data are substituted with their values.
```
Comment (pyrocto) [inline rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: Maybe capitalize for consistency w/ the rest of your comments?

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+sealed trait Ground extends Expr
+case class EList(ps: List[Par]) extends Ground
+case class ETuple(ps: List[Par]) extends Ground
+case class ESet(ps: List[Par]) extends Ground
+case class EMap(kvs: List[(Par,Par)]) extends Ground
+case class ENeg(p: Par) extends Expr
+// A variable used as a var should be bound in a process context, not a name
+// context. For example:
+// for (@x <- c1; @y <- c2) { z!(x + y) } is fine, but
+// for (x <- c1; y <- c2) { z!(x + y) } should raise an error.
```
Comment (pyrocto) [inline rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: I think the error here is because + is expecting two processes, not two names.  If f were expecting a name and a process, then `for (x <- c1; @y <- c2) { z!(f(x,y)) }` should work, right?

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala (hunk 1 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: Yes.

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala (hunk 4 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: Yes, but there are no such f's at the moment.

There's a test that covers this assertion.

File: rholang/src/main/scala/rholang/interpreter/RholangADT.scala (hunk 2 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/RholangADT.scala]: When I get to this part in the normalizer, I'll write a test.

**Comment by guardbotmk3** [comment]
OK, review comments handled. Please re-approve.
